### ISSUE TITLE
sourceset output.dir() does not work with java-library plugin

### ISSUE THREAD

**Issue by mkmaier** [gradle/gradle#3231]
### Expected Behavior
In a multi-module build, adding an output directory to the main sourceSet (as explained in [SourceSetOutput](https://docs.gradle.org/current/dsl/org.gradle.api.tasks.SourceSetOutput.html)) should work the same when using the java plugin or the java-library plugin.

### Current Behavior
If the project in question is using the java-library plugin, only its classes directory is on the consumer's classpath (as explained [in the docs](https://docs.gradle.org/current/userguide/java_library_plugin.html#sec:java_library_known_issues_memory)), but not any extra output directories registered on the main sourceSet. This might also affect resources, since the resources directory is also not on the classpath of consumers.

**Comment by bmuschko** [comment]
Could you please provide a code example that reproduces the issue?

**Comment by mkmaier** [comment]
Of course. Please find it here: https://github.com/mkmaier/gradle-3231

**Comment by bmuschko** [comment]
Is there any reason why you don't model the `main` source set like this?

```
sourceSets {
    main {
        java {
            srcDirs = ['src/main/java', 'src/second/java']
        }
    }
}
```

I saw that you use directories like "generated" in your sample though no code is actually generated. Is that relevant? Is code generated and then supposed to be on the compile classpath for the main source set? Please provide the relevant information and the code that you use to generate code (if that's what you are trying to do).

**Comment by mkmaier** [comment]
This is only a reproducer mostly taken from gradle documentation examples. I tried to keep things as minimal as possible to highlight what I think is a bug, without going into details of why we run into this. The salient point for me is that it does work with the 'java' plugin, but this might be a conscious design choice. If so, I did not find it documented.

**Comment by mkmaier** [comment]
But since you asked, this is what we are really doing that runs into this problem. I'd welcome any suggestions for improvement as this also has the downside of not compiling in either IDEA or Eclipse.

The gist of what we do is:
1. compile the sources as `intermediate`
1. generate java sources from the intermediate classes (task `createWrappers`)
1. (copy and) modify the intermediate classes (task `wrap`)
1. compile the generated sources from 2. and the add the modified classes from 3. as output in the `main` sourceSet (the unmodified intermediate classes are not part of this)

I hope this is clear enough, but I cannot provide much more details, especially not for the two ant tasks.

The relevant lines are these:


    // generated java source
    def srcGen = file("${buildDir}/src-gen")
    
    // generated binaries
    def wrappers = file("${buildDir}/wrappers")
    
    sourceSets {
        // original sources, not directly part of output
        intermediate {
            java.srcDirs = ['src']
        }
        // generated sources plus modified classes from original sources
        main {
            java.srcDirs = [srcGen]
            output.dir(wrappers, builtBy: 'wrap')
            compileClasspath += sourceSets.intermediate.output
        }
    }
    
    // generate modified classes
    task wrap() {
        inputs.files compileIntermediateJava
        outputs.dir wrappers
    
        doLast {
            ant.wrap(srcdir: sourceSets.intermediate.java.outputDir, targetdir: wrappers)
        }
    }
    
    // generate sources
    task createWrappers() {
        inputs.files compileIntermediateJava
        outputs.dir srcGen
    
        doLast {
            ant.createwrappers(srcdir: sourceSets.intermediate.java.outputDir, targetdir: srcGen)
        }
    }
    tasks.compileJava.dependsOn createWrappers

**Comment by mkmaier** [comment]
I have the same problem in another subproject as well, now, where we need to compile parts of the code with an older target release:

    apply plugin: 'java-library'
	sourceSets {
		legacy {
			java {
				srcDirs = ['src']
				include 'com/acme/legacy/**'
				compileClasspath += sourceSets.main.compileClasspath
			}
		}
		main {
			java {
				srcDirs = ['src']
				exclude 'com/acme/legacy/**'
				compileClasspath += sourceSets.legacy.output
				output.dir(sourceSets.legacy.output, builtBy: tasks[sourceSets.legacy.getCompileTaskName('java')])
			}
		}
	}
    tasks[sourceSets.legacy.getCompileTaskName('java')].sourceCompatibility = 1.6
    tasks[sourceSets.legacy.getCompileTaskName('java')].targetCompatibility = 1.6

If I use the 'java-library' plugin, the consumers cannot see the legacy classes and fail to compile. Without the second sourceSet, or with the 'java' plugin, all is well. Do you agree that this is a bug, or is this behaviour intended?

**Comment by alllex** [comment]
Sorry for the late reply.

The issue is in the backlog of the relevant team and is prioritized by them.
***

This is still reproducible with the latest released Gradle 9.5.1
